File: LeetCodeProblemsGo/LeetCode1461CheckIfAStringContainsAllBinaryCodesOfSizeK/LeetCode1461CheckIfAStringContainsAllBinaryCodesOfSizeK.go
package LeetCode1461CheckIfAStringContainsAllBinaryCodesOfSizeK

import "fmt"

func hasAllCodes(s string, k int) bool {
	length := len(s)
	if length < k {
		return false
	}
	nums := make(map[int]any)
	candidate := 0
	for i := 0; i < k; i++ {
		candidate = candidate << 1
		if s[i] == '1' {
			candidate++
		}
	}
	nums[candidate] = nil
	for i := k; i < length; i++ {
		candidate &= ^(1 << (k - 1))
		candidate <<= 1
		if s[i] == '1' {
			candidate++
		}
		nums[candidate] = nil
	}
	return len(nums) == 1<<k
}
func Calculate() {
	k := 2
	s := "00110110"
	fmt.Println(hasAllCodes(s, k))
}

File: LeetCodeProblemsCMake/Problems/LeetCode1461CheckIfAStringContainsAllBinaryCodesOfSizeK.cpp
//
// Created by Saeed Jannati on 2/23/2026 A.
//

#include "LeetCode1461CheckIfAStringContainsAllBinaryCodesOfSizeK.h"

#include <unordered_map>
#include <unordered_set>
using namespace std;

bool LeetCode1461CheckIfAStringContainsAllBinaryCodesOfSizeK::hasAllCodes(string s, int k) {
    int length = static_cast<int>(s.length());
    if (length - k + 1 < 1 << k) {
        return false;
    }
    unordered_set<int> nums{};
    int candidate = 0;
    int target = 1 << k;
    for (int i = 0; i < k; i++) {
        candidate = (candidate << 1) + (s[i] == '1' ? 1 : 0);
    }
    nums.insert(candidate);
    int mask = ~target;
    for (int i = k; i < length; i++) {
        candidate = ((candidate << 1) + (s[i] == '1' ? 1 : 0)) & mask;
        nums.insert(candidate);
    }
    return nums.size() == target;
}
